pkg/togroups/aws.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package togroups
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/external"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+const (
+	awsLogLevel            = aws.LogOff // For debugging pourposes can be set to aws.LogDebugWithSigning
+	AWS_DEFAULT_REGION_KEY = "AWS_DEFAULT_REGION"
+	AWS_DEFAULT_REGION     = "eu-west-1"
+)
+
+var (
+	POLICY_SECURITY_GROUP_ID_KEY = aws.String("instance.group-id")
+	POLICY_SECURITY_GROUP_NAME   = aws.String("instance.group-name")
+	POLICY_EC2_LABELS_KEY        = "tag"
+)
+
+func init() {
+	api.RegisterToGroupsProvider(api.AWSPROVIDER, GetIPsFromGroup)
+}
+
+// GetIpsFromGroup will return the list of the ips for the given group filter
+func GetIPsFromGroup(group *api.ToGroups) ([]net.IP, error) {
+	result := []net.IP{}
+	if group.Aws == nil {
+		return result, fmt.Errorf("no aws data available")
+	}
+	ips, err := GetInstancesIpsFromFilter(group.Aws)
+	return ips, err
+}
+
+// InitializeAWSAccount retrieve the env variables from the runtime and it
+// iniliazes the account in the specified region.
+func InitializeAWSAccount(region string) (*aws.Config, error) {
+	cfg, err := external.LoadDefaultAWSConfig()
+	if err != nil {
+		return nil, fmt.Errorf("Cannot initialize aws connector: %s", err)
+	}
+	cfg.Region = GetDefaultRegion()
+	cfg.LogLevel = awsLogLevel
+	return &cfg, nil
+}
+
+//GetInstancesFromFilter returns the instances IPs in aws EC2 filter by the
+//given filter
+func GetInstancesIpsFromFilter(filter *api.AWSGroups) ([]net.IP, error) {
```

Comment:
I think this will not scale well. The more CNPs you add, the more often you will have to query the metadata server. Can we fetch the entire list of IPs including labels instead and then filter on our own?